__tests__/unit/fetch-client.test.ts

```diff
@@ -0,0 +1,113 @@
+  read_ops: 0,
+  write_ops: 0,
+  query_time_ms: 0,
+  storage_bytes_read: 0,
+  storage_bytes_written: 0,
+  contention_retries: 0,
+};
+
+describe("query", () => {
+  beforeAll(() => {
+    fetchClient = new FetchClient();
+  });
+
+  beforeEach(() => {
+    fetchMock.resetMocks();
+  });
+
+  it("returns a valid query response on success", async () => {
+    expect.assertions(7);
+    const expected: QuerySuccess<null> = {
+      data: null,
+      summary: "summary",
+      txn_time: "2023-02-17T19:19:12.887330Z",
+      query_tags: {},
+      stats: dummyStats,
+    };
+    const expectedHeaders = {
+      "content-type": "	application/json;charset=utf-8",
+      "x-faunadb-build": "230217.014809-a23359a",
+      "x-some-response-header": "Some header value",
+    };
+    fetchMock.mockResponseOnce(JSON.stringify(expected), {
+      headers: expectedHeaders,
+    });
+    const response: HTTPResponse = await fetchClient.request(dummyRequest);
+    if (isHTTPResponse(response)) {
+      expect(response.status).toEqual(200);
+      expect(response.headers).toEqual(expectedHeaders);
+      const body = JSON.parse(response.body);
+      expect(body.data).toEqual(expected.data);
+      expect(body.summary).toEqual(expected.summary);
+      expect(body.txn_time).toEqual(expected.txn_time);
+      expect(body.query_tags).toEqual(expected.query_tags);
+      expect(body.stats).toEqual(expected.stats);
+    }
+  });
+
+  it("returns a valid query response on failure", async () => {
+    expect.assertions(7);
+    const expected: QueryFailure = {
+      error: {
+        code: "invalid_query",
+        message: "The query failed 1 validation check",
+      },
+      summary: "summary",
+      txn_time: "2023-02-17T19:19:12.887330Z",
+      query_tags: {},
+      stats: dummyStats,
+    };
+    const expectedHeaders = {
+      "content-type": "	application/json;charset=utf-8",
+      "x-faunadb-build": "230217.014809-a23359a",
+      "x-some-response-header": "Some header value",
+    };
+    fetchMock.mockResponseOnce(JSON.stringify(expected), {
+      status: 400,
+      headers: expectedHeaders,
+    });
+    const response: HTTPResponse = await fetchClient.request(dummyRequest);
+    if (isHTTPResponse(response)) {
+      expect(response.status).toEqual(400);
+      expect(response.headers).toEqual(expectedHeaders);
+      const body = JSON.parse(response.body);
+      expect(body.error).toEqual(expected.error);
+      expect(body.summary).toEqual(expected.summary);
+      expect(body.txn_time).toEqual(expected.txn_time);
+      expect(body.query_tags).toEqual(expected.query_tags);
+      expect(body.stats).toEqual(expected.stats);
+    }
+  });
+
+  it("forwards errors when fetch rejects", async () => {
+    expect.assertions(1);
+    fetchMock.mockRejectOnce(new Error("oops"));
+    try {
+      await fetchClient.request(dummyRequest);
+    } catch (error) {
+      expect((error as Error).message).toEqual("oops");
+    }
+  });
```

Comment:
This has me thinking about error handling.

Right now the FetchClient just passes any errors on to the Client to deal with. This is moderately okay, because the implementation is so simple that pretty much the only real place it can fail is if there is if `fetch` rejects. The Client can assume any exceptions thrown are NetworkErrors.

I don't know if that is going to hold for other implementations. It seems likely that we will handle specific error cases, for example, with the Node HTTP2 implementation.

We probably want to enforce some contract that HTTPClients have a common base error to throw. We have several classes that extend the `Error` class. But we could create our own abstract `FaunaError` class from which all of our errors derive.

All that said, I think this test is reasonable, because all of our error classes make sure to construct the base Error class with the original failure message. E.g. we could catch and rethrow a different Error from the FetchClient and the test aught to continue to pass. We can add another test to check for the common Error class if we want.

Reply:
#64 provides that common error class. With that, I could update the fetch implementation to handle exceptions from `fetch` and add a test that it always returns an instance of the base class.

Reply:
@ptpaterson let's punt this for now.

For discussion, I don't think we'd want the contract to be it throws a top-level error that is the parent of all errors - becuase if we end up making that error type we should make it abstract (that is never instantiate an actual member of it) because it doesn't really convey any information about what failed.

The concern would be that other HTTPClient implementations could get sloppy and emit a concrete version of an error that doesn't tell us what actually failed.

That being, said working on a contract for the the HTTP layer's request that includes what errors it fires sounds good - but let's punt it for the time being. I suspect it will be easier to work on a sane common contract once we have more than a single implementation.